File: firebase/functions/lib/utils/infra/firestore_utils.dart
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:data_models/utils/utils_web.dart';

/// Singleton firebase app that is set by main().
///
/// We usually use getit, but it was failing due to null safety issues.
late App Function() _firebaseAppFactory;
App? _firebaseAppSingleton;

void setFirebaseAppFactory(App Function() factory) {
  _firebaseAppFactory = factory;
}

App get firebaseApp => _firebaseAppSingleton ??= _firebaseAppFactory();
Firestore get firestore => firebaseApp.firestore();

/// Use a global singleton to allow mocking in tests
FirestoreUtils firestoreUtils = FirestoreUtils();

class FirestoreUtils {
  Future<T> getFirestoreObject<T>({
    required String path,
    required T Function(Map<String, dynamic>) constructor,
    Transaction? transaction,
  }) async {
    final ref = firestore.document(path);
    final snapshot =
        transaction == null ? await ref.get() : await transaction.get(ref);
    return constructor(fromFirestoreJson(snapshot.data.toMap()));
  }

  Map<String, dynamic> fromFirestoreJson(Map<String, dynamic> json) {
    var formattedJson = <String, dynamic>{};

    json.forEach((key, value) {
      if (value is Timestamp) {
        formattedJson[key] = value.toDateTime();
      } else if (value is Map) {
        formattedJson[key] = fromFirestoreJson(value as Map<String, dynamic>);
      } else if (value is List) {
        formattedJson[key] = value.map((v) {
          if (v is Map) {
            return fromFirestoreJson(v as Map<String, dynamic>);
          } else {
            return v;
          }
        }).toList();
      } else {
        formattedJson[key] = value;
      }
    });

    return formattedJson;
  }

  Map<String, dynamic> toFirestoreJson(Map<String, dynamic> json) {
    var formattedJson = <String, dynamic>{};

    json.forEach((key, value) {
      if (value == serverTimestampValue) {
        formattedJson[key] = Firestore.fieldValues.serverTimestamp();
      }
      // Check if the value is a DateTime object
      else if (value is DateTime) {
        formattedJson[key] = Timestamp.fromDateTime(value);
      } else if (value is Map) {
        formattedJson[key] = toFirestoreJson(value as Map<String, dynamic>);
      } else if (value is List) {
        formattedJson[key] = value.map((v) {
          if (v is Map) {
            return toFirestoreJson(v as Map<String, dynamic>);
          } else {
            return v;
          }
        }).toList();
      } else {
        formattedJson[key] = value;
      }
    });

    return formattedJson;
  }
}

extension DocumentQueryExtended on DocumentQuery {
  DocumentQuery whereNotEqual(String field, {String? notEqualTo}) {
    // ignore: invalid_use_of_protected_member
    final query = nativeInstance.where(field, '!=', notEqualTo);
    return DocumentQuery(query, firestore);
  }
}

File: firebase/functions/src/utils/infra/firestore_utils.ts
import * as admin from 'firebase-admin'

// Initialize Firebase Admin SDK (only once)
if (!admin.apps.length) {
    admin.initializeApp()
}

export const firebaseApp = admin.app()
export const firestore = admin.firestore()

export const isEmulator = process.env.FUNCTIONS_EMULATOR === 'true'

export class FirestoreUtils {
    async getFirestoreObject<T>({
        path,
        constructor,
        transaction,
    }: {
        path: string
        constructor: (map: Record<string, unknown>) => T
        transaction?: admin.firestore.Transaction
    }): Promise<T> {
        const ref = firestore.doc(path)
        const snapshot = transaction ? await transaction.get(ref) : await ref.get()
        const map = this.fromFirestoreJson(snapshot.data() ?? {})
        // Populate `id` from the document reference if missing or empty
        if (!map['id'] || typeof map['id'] !== 'string' || (map['id'] as string).length === 0) {
            map['id'] = ref.id
        }
        return constructor(map)
    }

    fromFirestoreJson(json: Record<string, unknown>): Record<string, unknown> {
        const formatted: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(json)) {
            if (value instanceof admin.firestore.Timestamp) {
                formatted[key] = value.toDate()
            } else if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
                formatted[key] = this.fromFirestoreJson(value as Record<string, unknown>)
            } else if (Array.isArray(value)) {
                formatted[key] = value.map((v) =>
                    v !== null && typeof v === 'object' && !Array.isArray(v)
                        ? this.fromFirestoreJson(v as Record<string, unknown>)
                        : v
                )
            } else {
                formatted[key] = value
            }
        }
        return formatted
    }

    toFirestoreJson(json: Record<string, unknown>): Record<string, unknown> {
        const formatted: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(json)) {
            if (value === 'SERVER_TIMESTAMP') {
                formatted[key] = admin.firestore.FieldValue.serverTimestamp()
            } else if (value instanceof Date) {
                formatted[key] = admin.firestore.Timestamp.fromDate(value)
            } else if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
                formatted[key] = this.toFirestoreJson(value as Record<string, unknown>)
            } else if (Array.isArray(value)) {
                formatted[key] = value.map((v) =>
                    v !== null && typeof v === 'object' && !Array.isArray(v)
                        ? this.toFirestoreJson(v as Record<string, unknown>)
                        : v
                )
            } else {
                formatted[key] = value
            }
        }
        return formatted
    }
}

// Global singleton — can be replaced in tests
export let firestoreUtils = new FirestoreUtils()
export function setFirestoreUtils(instance: FirestoreUtils): void {
    firestoreUtils = instance
}
